src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/osd/osd-recv-speed-modal/osd-recv-speed-modal.component.html

```diff
@@ -16,7 +16,8 @@
             <select class="form-control custom-select"
                     formControlName="priority"
                     id="priority"
-                    (change)="onPriorityChange($event.target.value)">
+                    (change)="onPriorityChange($event.target.value)"
+                    cdDisableEdit="osd">
```

Comment:
Suggestion: instead of defining multiple times the "osd" scope, you could define at a higher level a new directive `cdScope="osd"` and inject that scope in the Directive constructor from [Hierarchical Element DI](https://angular.io/guide/hierarchical-dependency-injection#directive-and-component):
```ts
  constructor(
      public scope: ScopeDirective,
      private authStorageService: AuthStorageService) {
    this.permission = this.authStorageService.getPermissions()[scope.scope];
  }
```

Alternatively, and as I wrote this I leaned towards this approach (because it allows to do all kind of UI customizations based on permissions), if you have such a scope directive, instead of having a specific cdDisableEdit Directive you could use [template references](https://angular.io/guide/template-syntax#template-reference-variables-var) to access that scope directive and:
```ts
@Directive({
  selector: '[cdScope]',
  exportAs: 'scope'
})
export class ScopeDirective {
  private permission: Permission;

  @Input('cdScopeDirective') scope: keyof Permissions;

  constructor(private authStorageService: AuthStorageService) {
    this.permission = this.authStorageService.getPermissions()[this.scope];
  }
}
```

```html
<form cdScope="osd" #thisScope="scope" ...>
  <select [disable]="thisScope.permission.edit" ...>
  <span [ngStyle]="{'background-colour': thisScope.permission.delete ? 'red' : 'gray'}"
</form>
```

Reply:
Cool. I'll work on to bring up something like this. Will reach out to you incase I need some clarification. Thanks

Reply:
> Suggestion: instead of defining multiple times the "osd" scope, you could define at a higher level a new directive `cdScope="osd"` and inject that scope in the Directive constructor from [Hierarchical Element DI](https://angular.io/guide/hierarchical-dependency-injection#directive-and-component):
> 
> ```ts
>   constructor(
>       public scope: ScopeDirective,
>       private authStorageService: AuthStorageService) {
>     this.permission = this.authStorageService.getPermissions()[scope.scope];
>   }
> ```
> 
> Alternatively, and as I wrote this I leaned towards this approach (because it allows to do all kind of UI customizations based on permissions), if you have such a scope directive, instead of having a specific cdDisableEdit Directive you could use [template references](https://angular.io/guide/template-syntax#template-reference-variables-var) to access that scope directive and:
> 
> ```ts
> @Directive({
>   selector: '[cdScope]',
>   exportAs: 'scope'
> })
> export class ScopeDirective {
>   private permission: Permission;
> 
>   @Input('cdScopeDirective') scope: keyof Permissions;
> 
>   constructor(private authStorageService: AuthStorageService) {
>     this.permission = this.authStorageService.getPermissions()[this.scope];
>   }
> }
> ```
> 
> ```
> <form cdScope="osd" #thisScope="scope" ...>
>   <select [disable]="thisScope.permission.edit" ...>
>   <span [ngStyle]="{'background-colour': thisScope.permission.delete ? 'red' : 'gray'}"
> </form>
> ```

Having this kind of implementation sounds overall great. One thing I noticed is that while disabling the `Cluster-wide OSD Flags`, there are 4 flags which are disabled everytime regardless of the permissions. These are some exception cases which I am entirely not sure how to be dealt with the way you proposed. 

Also, while making changes to the config-option.component, I noticed that I was not able to bind the data with `[cdDisableEdit]` or similar commands. The make check always fails if I use like this. So there are also these types of situation which I am not entirely sure how to deal with.

Reply:
Reviewing my last proposal, I think you can go one step forward and just only need a single explicit directive: if you create a high-level directive (e.g.: `cdScope`, `cdPermissionScope` or `cdFeature`), instead of needing to apply lower-level directives for handling specific behaviours, there are ways you can alter 'children' behaviours from that higher-level one.

Directive selectors not always have to be attribute selectors or they can be applied to standard HTML elements (e.g.: [ngForm](https://github.com/angular/angular/blob/master/packages/forms/src/directives/ng_form.ts) selector is applied by default to all forms except those with `ngNoForm` attribute: `selector: 'form:not([ngNoForm]):not([formGroup]),ng-form,[ngForm]'`).

You could take a similar approach here:
* Either by using `@ContentChildred` for accessing children elements (e.g.: select) and enabling/disabling those from the parent `cdScope` element.
* Or by creating (implicit) directives for subelements (e.g.: `selector: select` doesn't need HTML template modifications) and then getting references to the parent `cdScope` via Dependency Injection from the constructor.

Wish it'd be clear, while I'm not that sure. Let me know if you need any help with this.

Reply:
> Having this kind of implementation sounds overall great. One thing I noticed is that while disabling the Cluster-wide OSD Flags, there are 4 flags which are disabled everytime regardless of the permissions. These are some exception cases which I am entirely not sure how to be dealt with the way you proposed.

And answering specifically to your question: you can always define a negative matching selector (same way as the ngNoForm) for excluding those elements from the default behaviour.